File: java/1046.last-stone-weight.java
/*
 * @lc app=leetcode id=1046 lang=java
 *
 * [1046] Last Stone Weight
 */

// @lc code=start
class Solution {
    public int lastStoneWeight(int[] stones) {
        PriorityQueue<Integer> pq = new PriorityQueue(Comparator.reverseOrder());

        for(int stone: stones) {
            pq.add(stone);
        }

        while(pq.size() > 1) {
            int x = pq.poll();
            int y = pq.poll();

            if(x > y) {
                pq.add(x - y);
            }
        }

        return pq.size() == 0? 0 : pq.poll();
    }
}
// @lc code=end

File: cpp/1046.last-stone-weight.cpp
/*
 * @lc app=leetcode id=1046 lang=cpp
 *
 * [1046] Last Stone Weight
 */

// @lc code=start
class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        priority_queue<int> pq {stones.begin(), stones.end()};

        int x, y;
        while(pq.size() > 1) {
            y = pq.top();
            pq.pop();
            x = pq.top();
            pq.pop();

            if(x != y) {
                pq.push(y - x);
            }
        }

        return pq.size() == 0? 0 : pq.top();
    }
};
// @lc code=end

